docs/debugging/xl-meta/main.go

```diff
@@ -976,179 +1012,204 @@ func combine(files []string, out string) error {
+					// If data+parity matches, combine.
+					if m.parity == otherM.parity && m.data == otherM.data {
+						for k, v := range m.parityData {
+							if otherM.parityData[k] == nil {
 								continue
 							}
-							m.parityData[k][i] = otherM.parityData[k][i]
+							for i, data := range v {
+								if data != nil || otherM.parityData[k][i] == nil {
+									continue
+								}
+								m.parityData[k][i] = otherM.parityData[k][i]
+							}
 						}
 					}
-				}
 
-				fmt.Printf("Data overlaps (%d bytes). Combining with %q.\n", ok, names[j])
-				for i := range otherM.filled {
-					if otherM.filled[i] == 1 {
-						m.filled[i] = 1
-						m.mapped[i] = otherM.mapped[i]
+					fmt.Printf("Data overlaps (%d bytes). Combining with %q.\n", ok, otherKey)
+					for i := range otherM.filled {
+						if otherM.filled[i] == 1 {
+							m.filled[i] = 1
+							m.mapped[i] = otherM.mapped[i]
+						}
 					}
 				}
-			}
-			lastValid := 0
-			missing := 0
-			for i := range m.filled {
-				if m.filled[i] == 1 {
-					lastValid = i
-				} else {
-					missing++
-				}
-			}
-			if missing > 0 && len(m.parityData) > 0 {
-				fmt.Println("Attempting to reconstruct using parity sets:")
-				for k, v := range m.parityData {
-					if missing == 0 {
-						break
-					}
-					fmt.Println("* Setup: Data shards:", k, "- Parity blocks:", m.shards-k)
-					rs, err := reedsolomon.New(k, m.shards-k)
-					if err != nil {
-						return err
-					}
-					split, err := rs.Split(m.mapped)
-					if err != nil {
-						return err
-					}
-					splitFilled, err := rs.Split(m.filled)
-					if err != nil {
-						return err
+				lastValid := 0
+				missing := 0
+				for i := range m.filled {
+					if m.filled[i] == 1 {
+						lastValid = i
+					} else {
+						missing++
 					}
-					ok := len(splitFilled)
-					for i, sh := range splitFilled {
-						for _, v := range sh {
-							if v == 0 {
-								split[i] = nil
-								ok--
-								break
+				}
+				if missing > 0 && len(m.parityData) > 0 {
+					fmt.Println("Attempting to reconstruct using parity sets:")
+					for k, v := range m.parityData {
+						if missing == 0 {
+							break
+						}
+						fmt.Println("* Setup: Data shards:", k, "- Parity blocks:", m.shards-k)
+						rs, err := reedsolomon.New(k, m.shards-k)
+						if err != nil {
+							return err
+						}
+						split, err := rs.Split(m.mapped)
+						if err != nil {
+							return err
+						}
+						splitFilled, err := rs.Split(m.filled)
+						if err != nil {
+							return err
+						}
+						ok := len(splitFilled)
+						for i, sh := range splitFilled {
+							for _, v := range sh {
+								if v == 0 {
+									split[i] = nil
+									ok--
+									break
+								}
 							}
 						}
-					}
-					hasParity := 0
-					for idx, sh := range v {
-						split[idx] = sh
-						if idx >= k && len(v) > 0 {
-							hasParity++
+						hasParity := 0
+						for idx, sh := range v {
+							split[idx] = sh
+							if idx >= k && len(v) > 0 {
```

Comment:
Shouldn't we be checking if the shard `sh` at drive `idx` is present?
```suggestion
							if idx >= k && len(sh) > 0 {
```

Reply:
@klauspost ^^

Reply:
Would look like it. But I think it is just a needless secondary check, since we on have valid parity shards here anyway.